dygraph/infer.py

```diff
@@ -0,0 +1,174 @@
+        "--input_size",
+        dest="input_size",
+        help="The image size for net inputs.",
+        nargs=2,
+        default=[512, 512],
+        type=int)
+    parser.add_argument(
+        '--batch_size',
+        dest='batch_size',
+        help='Mini batch size',
+        type=int,
+        default=2)
+    parser.add_argument(
+        '--model_dir',
+        dest='model_dir',
+        help='The path of model for evaluation',
+        type=str,
+        default=None)
+    parser.add_argument(
+        '--save_dir',
+        dest='save_dir',
+        help='The directory for saving the inference results',
+        type=str,
+        default='./output/result')
+
+    return parser.parse_args()
+
+
+def mkdir(path):
+    sub_dir = osp.dirname(path)
+    if not osp.exists(sub_dir):
+        os.makedirs(sub_dir)
+
+
+def infer(model, data_dir=None, test_list=None, model_dir=None,
+          transforms=None):
+    ckpt_path = osp.join(model_dir, 'model')
+    para_state_dict, opti_state_dict = fluid.load_dygraph(ckpt_path)
+    model.set_dict(para_state_dict)
+    model.eval()
+
+    added_saved_dir = osp.join(args.save_dir, 'added')
+    pred_saved_dir = osp.join(args.save_dir, 'prediction')
+
+    logging.info("Start to predict...")
+    with open(test_list, 'r') as f:
+        files = f.readlines()
+    for file in tqdm.tqdm(files):
+        file = file.strip()
+        im_file = osp.join(data_dir, file)
+        im, im_info = transforms(im_file)
+        im = np.expand_dims(im, axis=0)
+        im = to_variable(im)
+
+        pred, _ = model(im, mode='test')
+        pred = pred.numpy()
+        pred = np.squeeze(pred).astype('uint8')
+        keys = list(im_info.keys())
+        for k in keys[::-1]:
+            if k == 'shape_before_resize':
+                h, w = im_info[k][0], im_info[k][1]
+                pred = cv2.resize(pred, (w, h), cv2.INTER_NEAREST)
+            elif k == 'shape_before_padding':
+                h, w = im_info[k][0], im_info[k][1]
+                pred = pred[0:h, 0:w]
+
+        # save added image
+        added_image = utils.visualize(im_file, pred, weight=0.6)
+        added_image_path = osp.join(added_saved_dir, file)
+        mkdir(added_image_path)
+        cv2.imwrite(added_image_path, added_image)
+
+        # save prediction
+        pred_im = utils.visualize(im_file, pred, weight=0.0)
+        pred_saved_path = osp.join(pred_saved_dir, file)
+        mkdir(pred_saved_path)
+        cv2.imwrite(pred_saved_path, pred_im)
+
+
+def arrange_transform(transforms, mode='train'):
+    arrange_transform = T.ArrangeSegmenter
+    if type(transforms.transforms[-1]).__name__.startswith('Arrange'):
+        transforms.transforms[-1] = arrange_transform(mode=mode)
+    else:
+        transforms.transforms.append(arrange_transform(mode=mode))
+
+
+def main(args):
+    test_transforms = T.Compose([T.Resize(args.input_size), T.Normalize()])
+    arrange_transform(test_transforms, mode='test')
```

Comment:
It's very weird here, is it possible not use use ArrangeSegmenter here? Since it's not easy to understand here.

Reply:
Has removed ArrangeSegmenter